### PR TITLE
apps/ca.c: Handle EVP_PKEY_get_default_digest_name() returning 1 with "UNDEF"

### PR DESCRIPTION
**Pull request by levitte**
`EVP_PKEY_get_default_digest_name()` may return 1 with the returned digest
name `"UNDEF"`.  This case hasn't been documented, and the meaning has been
left undefined, until now.

Related to #20428

### CODE REVIEW

**Comment by levitte** [comment]
I felt a little unsure if this should go to 3.0, even though #20430 did...

**Comment by levitte** [comment]
@baentsch, please have a look here

File: doc/man3/EVP_PKEY_get_default_digest_nid.pod [hunk 1]
```diff
@@ -18,8 +18,8 @@ EVP_PKEY_get_default_digest_nid, EVP_PKEY_get_default_digest_name
 EVP_PKEY_get_default_digest_name() fills in the default message digest
 name for the public key signature operations associated with key
 I<pkey> into I<mdname>, up to at most I<mdname_sz> bytes including the
-ending NUL byte.  The name could be C<"UNDEF">, signifying that no digest
-should be used.
+ending NUL byte.  The name could be C<"UNDEF">, signifying that a digest
+must (for return value 2) or may (for return value 1) be left unspecified.
```
Comment (levitte) [inline doc/man3/EVP_PKEY_get_default_digest_nid.pod:22]: Strictly speaking, this was possible with `EVP_PKEY_get_default_digest_nid()` as well.  No `EVP_PKEY_ASN1_METHOD` that I know actually did so, but it's perfectly possible to have the ctrl function to `*pnid = NID_undef` and return 1.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Should there be a test?

If it can be treated as bug fix, then it should go to 3.0 as well.

File: apps/ca.c [hunk 2]
```diff
@@ -806,15 +806,20 @@ int ca_main(int argc, char **argv)
     /*
      * EVP_PKEY_get_default_digest_name() returns 2 if the digest is
      * mandatory for this algorithm.
+     *
+     * That call may give back the name "UNDEF", which has these meanings:
+     *
+     * when def_ret == 2: the user MUST NOT specify a digest
+     * when def_ret == 1: the user MAY not specify a digest
```
Comment (baentsch) [inline apps/ca.c:813]: I'm no native English speaker, so please take this feedback with a grain of salt: Should this comment not better use the word "MIGHT" instead of "MAY"? "...may not specify..." sounds like a mandate; "...might not specify..." more like the option (it is meant to be).

File: doc/man3/EVP_PKEY_get_default_digest_nid.pod (hunk 1 above)
Comment (baentsch) [inline doc/man3/EVP_PKEY_get_default_digest_nid.pod:22]: In this context, the "optional character" of the (mandatory digest) parameter becomes clear (for ret val 1). But then again, there may be people only reading the code and not the documentation ... :-)

**Comment by baentsch** [comment]
> @baentsch, please have a look here

Just did. What I tried to test this out was to return _only_ "OSSL_PKEY_PARAM_DEFAULT_DIGEST" from oqsprovider (value "UNDEF"). This now passes the previously "suboptimal" checks at https://github.com/openssl/openssl/blob/931369429564b5a9bb09711de8e885fef546a0ac/apps/ca.c#L810-L824 updated via this PR (good!) but unfortunately now triggers a problem at https://github.com/openssl/openssl/blob/931369429564b5a9bb09711de8e885fef546a0ac/crypto/evp/m_sigver.c#L205-L214: (bad): It is now possible (don't know exactly why, though) to hit this label with an mdname="UNDEF". Patching this as follows:

```
diff --git a/crypto/evp/m_sigver.c b/crypto/evp/m_sigver.c
index 630d339c35..3286826ba9 100644
--- a/crypto/evp/m_sigver.c
+++ b/crypto/evp/m_sigver.c
@@ -206,6 +206,7 @@ static int do_sigver_init(EVP_MD_CTX *ctx, EVP_PKEY_CTX **pctx,
     if (pctx != NULL)
         *pctx = locpctx;
 
+    mdname = canon_mdname(mdname);
     if (type != NULL) {
         ctx->reqdigest = type;
         if (mdname == NULL)
@@ -218,7 +219,6 @@ static int do_sigver_init(EVP_MD_CTX *ctx, EVP_PKEY_CTX **pctx,
                 mdname = canon_mdname(locmdname);
             }
         }
```

has everything working fine (an `openssl ca` test built into `oqsprovider`).

**Comment by levitte** [comment]
> ``` diff
> diff --git a/crypto/evp/m_sigver.c b/crypto/evp/m_sigver.c
> index 630d339c35..3286826ba9 100644
> --- a/crypto/evp/m_sigver.c
> +++ b/crypto/evp/m_sigver.c
> @@ -206,6 +206,7 @@ static int do_sigver_init(EVP_MD_CTX *ctx, EVP_PKEY_CTX **pctx,
>      if (pctx != NULL)
>          *pctx = locpctx;
>  
> +    mdname = canon_mdname(mdname);
>      if (type != NULL) {
>          ctx->reqdigest = type;
>          if (mdname == NULL)
> @@ -218,7 +219,6 @@ static int do_sigver_init(EVP_MD_CTX *ctx, EVP_PKEY_CTX **pctx,
>                  mdname = canon_mdname(locmdname);
>              }
>          }
> ```

That looks a bit late...  how about this?

``` diff
diff --git a/crypto/evp/m_sigver.c b/crypto/evp/m_sigver.c
index 630d339c35..40312d4d79 100644
--- a/crypto/evp/m_sigver.c
+++ b/crypto/evp/m_sigver.c
@@ -88,6 +88,7 @@ static int do_sigver_init(EVP_MD_CTX *ctx, EVP_PKEY_CTX **pctx,
         goto err;
     }
 
+    mdname = canon_mdname(mdname);
     if (!reinit) {
         evp_pkey_ctx_free_old_ops(locpctx);
     } else {
```

File: apps/ca.c (hunk 2 above)
Comment (levitte) [reply apps/ca.c:813]: I hear you suggest that I reword it to be like the documentation.  Good suggestion, will do

**Comment by levitte** [comment]
> Should there be a test?

I was afraid you were going to ask this....  I'm afraid that writing a test for this is a *large* body of work.

**Comment by baentsch** [comment]
> That looks a bit late... how about this?

LGTM (and allows my CA test suite to pass).

**Comment by baentsch** [comment]
> I'm afraid that writing a test for this is a large body of work.

Completely concur. What could be called for are tests for all openssl commands utilizing providers -- times all their parameter options -- that's a true "test case explosion". The underlying issue for _this_ PR has been triggered by `openssl ca -provider ....`.

**Comment by t8m** [comment]
> > Should there be a test?
> 
> I was afraid you were going to ask this.... I'm afraid that writing a test for this is a _large_ body of work.

Then tests: exempted would apply.

**Comment by t8m** [comment]
> > ```diff
> > diff --git a/crypto/evp/m_sigver.c b/crypto/evp/m_sigver.c
> > index 630d339c35..3286826ba9 100644
> > --- a/crypto/evp/m_sigver.c
> > +++ b/crypto/evp/m_sigver.c
> > @@ -206,6 +206,7 @@ static int do_sigver_init(EVP_MD_CTX *ctx, EVP_PKEY_CTX **pctx,
> >      if (pctx != NULL)
> >          *pctx = locpctx;
> >  
> > +    mdname = canon_mdname(mdname);
> >      if (type != NULL) {
> >          ctx->reqdigest = type;
> >          if (mdname == NULL)
> > @@ -218,7 +219,6 @@ static int do_sigver_init(EVP_MD_CTX *ctx, EVP_PKEY_CTX **pctx,
> >                  mdname = canon_mdname(locmdname);
> >              }
> >          }
> > ```
> 
> That looks a bit late... how about this?
> 
> ```diff
> diff --git a/crypto/evp/m_sigver.c b/crypto/evp/m_sigver.c
> index 630d339c35..40312d4d79 100644
> --- a/crypto/evp/m_sigver.c
> +++ b/crypto/evp/m_sigver.c
> @@ -88,6 +88,7 @@ static int do_sigver_init(EVP_MD_CTX *ctx, EVP_PKEY_CTX **pctx,
>          goto err;
>      }
>  
> +    mdname = canon_mdname(mdname);
>      if (!reinit) {
>          evp_pkey_ctx_free_old_ops(locpctx);
>      } else {
> ```

The question is why the UNDEF mdname is passed to the do_sigver_init() at all.

**Comment by levitte** [comment]
> The question is why the UNDEF mdname is passed to the do_sigver_init() at all.

Someone may have received it by calling `EVP_PKEY_get_default_digest_name()` and passed it to any of the functions  `EVP_DigestSignInit_ex()`, `EVP_DigestSignInit()`, `EVP_DigestVerifyInit_ex()` or `EVP_DigestVerifyInit()`.

**Comment by t8m** [comment]
In that case I would consider the change suggested in https://github.com/openssl/openssl/pull/20460#issuecomment-1460272747 as a API convenience improvement.  However the question is why @baentsch sees it in his tests?

**Comment by levitte** [comment]
> In that case I would consider the change suggested in [#20460 (comment)](https://github.com/openssl/openssl/pull/20460#issuecomment-1460272747) as a API convenience improvement. However the question is why @baentsch sees it in his tests?

I don't understand this question.  `UNDEF` is a documented valid value under certain circumstances, just like `NID_undef` is a valid value when `EVP_PKEY_get_default_digest_nid()` is called, and @baentsch triggers exactly those circumstances...  except his specific circumstance was undocumented, because no one has thought of the case where the return value is 1 and the passed back digest nid/name is `NID_undef`/`UNDEF`...  except we do handle that in code:

https://github.com/openssl/openssl/blob/7eab7680ee61c64b2ae7acd9dd199ab6734f3d1f/crypto/evp/p_lib.c#L1332-L1338
https://github.com/openssl/openssl/blob/master/crypto/evp/keymgmt_lib.c#L560-L566

**Review by tmshort — APPROVED** [review]

**Comment by t8m** [comment]
So, is this a bug fix? If so IMO it should go to 3.0 too.

**Comment by paulidale** [comment]
I think it's a bug fix, if it isn't it can't go to 3.1.

**Comment by t8m** [comment]
Merged to master, 3.1, and 3.0 branches. Thank you.
